Batch up dynamic plugin installations from setup wizard

jglick opened the pull request:
Reduces CPU usage from the Setup Wizard when installing a batch of plugins. Not so dramatic an improvement in OSS Jenkins since the time is dominated by downloading plugins from a public update center; for CloudBees Jenkins variants, where the available plugins are bundled (`WEB-INF/plugins/*.hpi`), this speeds up installation by about 2×.

Most of the time in the unpatched installation is spent inside `refreshExtensions`, which is apparently pretty expensive. It does not suffice to just defer this call to the end, though: a plugin’s `@Initializer` generally expects its own extensions to be loaded, so all calls to plugin code also need to be deferred.

jglick committed:
Batch up dynamic plugin installations from setup wizard.

jglick commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -928,56 +933,67 @@ public void dynamicLoad(File arc, boolean removeExisting) throws IOException, In
+        }
+
+        // run initializers in the added plugins
+        Reactor r = new Reactor(InitMilestone.ordering());
```
Easier to read the diff if you ignore whitespace changes.

jglick committed:
Reworked to thread the list of batched-up plugins properly and introduced CompleteBatchJob to force the setup wizard to wait.

jglick commented:
Need to investigate failures (Java 11 only?):

```
java.lang.AssertionError: expected:<2> but was:<24>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at hudson.model.UpdateCenterPluginInstallTest.test_installKnownPlugins(UpdateCenterPluginInstallTest.java:84)
```

amuniz commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -928,56 +928,68 @@ public void dynamicLoad(File arc, boolean removeExisting) throws IOException, In
-                new InitReactorRunner().run(r);
-            } catch (ReactorException e) {
-                throw new IOException("Failed to initialize "+ sn +" plugin",e);
+            LOGGER.info("Plugin " + p.getShortName()+":"+p.getVersion() + " dynamically installed");
```
This log is misleading. It should mention whether the plugin installation has been batched or not.

jglick replied:
This log message is actually unchanged from before (surrounding code moved), but agreed that it would be better to say that it has been loaded but not started or something like that.

jglick committed:
Merge branch 'master' into PluginManager.install

jglick committed:
UpdateCenterPluginInstallTest.test_installKnownPlugins failure.

jglick committed:
@amuniz notes that an existing message could now be considered misleading.

Vlatombe commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -928,56 +928,68 @@ public void dynamicLoad(File arc, boolean removeExisting) throws IOException, In
-                new InitReactorRunner().run(r);
-            } catch (ReactorException e) {
-                throw new IOException("Failed to initialize "+ sn +" plugin",e);
+            LOGGER.log(INFO, "Plugin {0}:{1} dynamically {2}", new Object[] {p.getShortName(), p.getVersion(), batch != null ? "loaded but not yet started" : "installed"});
```
At the moment each plugin that gets installed gets 5 lines of logs.

```
juil. 24, 2019 3:08:24 PM hudson.model.UpdateCenter$DownloadJob run
INFOS: Starting the installation of Trilead API on behalf of anonymous
juil. 24, 2019 3:08:25 PM hudson.model.UpdateCenter$UpdateCenterConfiguration download
INFOS: Downloading Trilead API
juil. 24, 2019 3:08:27 PM hudson.PluginManager dynamicLoad
INFOS: Attempting to dynamic load /tmp/jenkins/war/work/plugins/trilead-api.jpi
juil. 24, 2019 3:08:27 PM hudson.PluginManager dynamicLoad
INFOS: Plugin trilead-api:1.0.4 dynamically loaded but not yet started
juil. 24, 2019 3:08:27 PM hudson.model.UpdateCenter$DownloadJob run
INFOS: Installation successful: Trilead API
```

I think this message should be on level `FINE`, as well as `Attempting to dynamic load....`. Additionally, I think the plugin display name, not the short name, should be used to be consistent with the rest of the plugin installation logging.

jglick replied:
> I think this message should be on level `FINE`, as well as `Attempting to dynamic load....`.

Probably a good idea.

> I think the plugin display name, not the short name, should be used to be consistent with the rest of the plugin installation logging.

To the contrary, I think short names should always be used in log files. Display names are for the GUI.

Vlatombe replied:
> To the contrary, I think short names should always be used in log files. Display names are for the GUI.

As long as it is consistent, I'm fine with it.

jglick commented on core/src/main/resources/hudson/model/UpdateCenter/CompleteBatchJob/row.jelly:
```diff
@@ -0,0 +1,10 @@
+  <tr id="row${it.id}">
+    <td style="vertical-align: top; padding-right:1em">${%Loading plugin extensions}</td>
+    <td style="vertical-align:middle">
+      <!-- TODO could set an actual status icon, if this is ever actually displayed in the UI (normally run from setup wizard which does not use this display) -->
```
OK, can try to fix that.

jglick committed:
Merge branch 'master' into PluginManager.install

jglick committed:
Introduced UpdateCenter.DownloadJob.getDisplayName for better consistency in logging, as suggested by @Vlatombe.

jglick committed:
Reducing log levels of messages from dynamicLoad which are not of great interest, as suggested by @Vlatombe.

jglick committed:
Reducing log level of another message of little interest.

jglick committed:
Adding classic UI for CompleteBatchJob.

jglick committed:
CompleteBatchJob should be used only when dynamicLoad.

jvz commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -928,56 +928,68 @@ public void dynamicLoad(File arc, boolean removeExisting) throws IOException, In
+
+    @Restricted(NoExternalUse.class)
+    public void start(List<PluginWrapper> plugins) throws Exception {
+        Jenkins.get().refreshExtensions();
```
I love it when refactoring finds things like this!

jvz reviewed (commented):
Code looks good, though I haven't tested it out yet. This feature sounds very useful for local testing purposes.

jglick committed:
Null defense on UpdateCenterJob.site.
Otherwise can get:
java.lang.NullPointerException
	at hudson.model.UpdateCenter.getConnectionCheckJob(UpdateCenter.java:850)
	at hudson.model.UpdateCenter.addConnectionCheckJob(UpdateCenter.java:817)
	at hudson.model.UpdateCenter.addJob(UpdateCenter.java:804)
	at hudson.model.UpdateSite$Plugin.deploy(UpdateSite.java:1367)
	at hudson.PluginManager.install(PluginManager.java:1537)
	at hudson.PluginManager.install(PluginManager.java:1496)
	at hudson.PluginManager.doInstall(PluginManager.java:1445)
	at …

Vlatombe commented:
Love this. We're down to 3 lines of logs per plugin when installing a batch of them.

```
juil. 25, 2019 8:56:52 AM hudson.model.UpdateCenter$DownloadJob run
INFOS: Starting the installation of trilead-api on behalf of anonymous
juil. 25, 2019 8:56:54 AM hudson.model.UpdateCenter$UpdateCenterConfiguration download
INFOS: Downloading trilead-api
juil. 25, 2019 8:56:55 AM hudson.model.UpdateCenter$DownloadJob run
INFOS: Installation successful: trilead-api
```

When installing a single plugin without any new dependency
```
juil. 25, 2019 9:00:34 AM hudson.PluginManager install
INFOS: Starting installation of a batch of 1 plugins plus their dependencies
juil. 25, 2019 9:00:34 AM hudson.model.UpdateCenter$DownloadJob run
INFOS: Starting the installation of build-monitor-plugin on behalf of anonymous
juil. 25, 2019 9:00:35 AM hudson.model.UpdateCenter$UpdateCenterConfiguration download
INFOS: Downloading build-monitor-plugin
juil. 25, 2019 9:00:36 AM hudson.model.UpdateCenter$DownloadJob run
INFOS: Installation successful: build-monitor-plugin
juil. 25, 2019 9:00:36 AM hudson.model.UpdateCenter$CompleteBatchJob run
INFOS: Completing installing of plugin batch…
juil. 25, 2019 9:00:36 AM jenkins.InitReactorRunner$1 onAttained
INFOS: Started initialization
juil. 25, 2019 9:00:36 AM jenkins.InitReactorRunner$1 onAttained
INFOS: Listed all plugins
juil. 25, 2019 9:00:36 AM jenkins.InitReactorRunner$1 onAttained
INFOS: Prepared all plugins
juil. 25, 2019 9:00:36 AM jenkins.InitReactorRunner$1 onAttained
INFOS: Started all plugins
juil. 25, 2019 9:00:36 AM jenkins.InitReactorRunner$1 onAttained
INFOS: Augmented all extensions
juil. 25, 2019 9:00:36 AM jenkins.InitReactorRunner$1 onAttained
INFOS: Loaded all jobs
juil. 25, 2019 9:00:36 AM jenkins.InitReactorRunner$1 onAttained
INFOS: Completed initialization
juil. 25, 2019 9:00:36 AM hudson.model.UpdateCenter$CompleteBatchJob run
INFOS: Completed installation of 1 plugins in 2,1 s

```

When uploading a plugin

```
juil. 25, 2019 9:04:49 AM hudson.model.UpdateCenter$DownloadJob run
INFOS: Starting the installation of ui-samples-plugin on behalf of anonymous
juil. 25, 2019 9:04:49 AM hudson.model.UpdateCenter$UpdateCenterConfiguration download
INFOS: Downloading ui-samples-plugin
juil. 25, 2019 9:04:50 AM hudson.ClassicPluginStrategy createClassJarFromWebInfClasses
AVERTISSEMENT: Created /tmp/jenkins/plugins/ui-samples-plugin/WEB-INF/lib/classes.jar; update plugin to a version created with a newer harness
juil. 25, 2019 9:04:50 AM jenkins.InitReactorRunner$1 onAttained
INFOS: Started initialization
juil. 25, 2019 9:04:50 AM jenkins.InitReactorRunner$1 onAttained
INFOS: Listed all plugins
juil. 25, 2019 9:04:50 AM jenkins.InitReactorRunner$1 onAttained
INFOS: Prepared all plugins
juil. 25, 2019 9:04:50 AM jenkins.InitReactorRunner$1 onAttained
INFOS: Started all plugins
juil. 25, 2019 9:04:50 AM jenkins.InitReactorRunner$1 onAttained
INFOS: Augmented all extensions
juil. 25, 2019 9:04:50 AM jenkins.InitReactorRunner$1 onAttained
INFOS: Loaded all jobs
juil. 25, 2019 9:04:50 AM jenkins.InitReactorRunner$1 onAttained
INFOS: Completed initialization
juil. 25, 2019 9:04:50 AM hudson.model.UpdateCenter$DownloadJob run
INFOS: Installation successful: ui-samples-plugin
```

> Most of the time in the unpatched installation is spent inside refreshExtensions, which is apparently pretty expensive.

Yes, this was of the downsides of #3370 (adding a new `refreshExtensions` call), and at the time, I wished plugin installation was batched. You made this wish come true, thanks!

oleg-nenashev commented:
CC @AbhyudayaSharma , could be a use-case for a new Jenkins core benchmark

oleg-nenashev commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -872,16 +872,16 @@ private boolean containsHpiJpi(Collection<String> bundledPlugins, String name) {
     /**
      * Try the dynamicLoad, removeExisting to attempt to dynamic load disabled plugins
      */
     @Restricted(NoExternalUse.class)
```
Note for reviewers. It was Restricted from very beginning, https://github.com/jenkinsci/jenkins/pull/2299

oleg-nenashev commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -872,16 +872,16 @@ private boolean containsHpiJpi(Collection<String> bundledPlugins, String name) {
+    public void dynamicLoad(File arc, boolean removeExisting, @CheckForNull List<PluginWrapper> batch) throws IOException, InterruptedException, RestartRequiredException {
         try (ACLContext context = ACL.as(ACL.SYSTEM)) {
-            LOGGER.info("Attempting to dynamic load "+arc);
+            LOGGER.log(FINE, "Attempting to dynamic load {0}", arc);
```
why unrelated?

Vlatombe replied:
The context is provided in earlier comments.

oleg-nenashev commented on core/src/main/java/hudson/model/UpdateCenter.java:
```diff
@@ -1492,7 +1495,7 @@ public ConnectionCheckJob(UpdateSite site) {
             connectionStates.put(ConnectionStatus.INTERNET, ConnectionStatus.UNCHECKED);
             connectionStates.put(ConnectionStatus.UPDATE_SITE, ConnectionStatus.UNCHECKED);
-            if (ID_UPLOAD.equals(site.getId())) {
+            if (site == null || ID_UPLOAD.equals(site.getId())) {
```
Does it deserve a log warning for the new check? Maybe also a separate PR?

jglick replied:
I do not think so. The NPE was observed only in a CloudBees update site configuration and is related to the new job type, which is not specific to a site.

oleg-nenashev commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -1334,18 +1334,19 @@ public void install() {
+                    dep.deploy(dynamicLoad, /* UpdateCenterPluginInstallTest.test_installKnownPlugins specifically asks that these not be correlated */ null, batch);
                 } else {
-                    LOGGER.log(Level.INFO, "Dependent install of " + dep.name + " for plugin " + name + " already added, skipping");
+                    LOGGER.log(Level.FINE, "Dependent install of {0} for plugin {1} already added, skipping", new Object[] {dep.name, name});
```
Why do you decrease logging level here?

Vlatombe replied:
ditto above

oleg-nenashev commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -1334,18 +1334,19 @@ public void install() {
          */
         @Restricted(NoExternalUse.class)
-        public Future<UpdateCenterJob> deploy(boolean dynamicLoad, @CheckForNull UUID correlationId) {
+        public Future<UpdateCenterJob> deploy(boolean dynamicLoad, @CheckForNull UUID correlationId, @CheckForNull List<PluginWrapper> batch) {
```
have you checked API usages there. It was not Restricted before https://github.com/jenkinsci/jenkins/pull/2269 by @daniel-beck

jglick replied:
From plugins built against Jenkins 1.x? That is pretty old. This is also not a sort of API generally expected to be used by plugins. AFAIK we have no tool to look for such method calls structurally (the method name is too generic to search for).

jglick replied:
```console
user@jenkinsci$ find *-plugin/src -type f -print0 | xargs -0 egrep -l 'UpdateSite|UpdateCenter' | xargs fgrep deploy
blackduck-installer-plugin/src/main/java/com/blackducksoftware/integration/jenkins/BDInstaller.java:                deployPlugin(vulnerabilityReportPlugin);
blackduck-installer-plugin/src/main/java/com/blackducksoftware/integration/jenkins/BDInstaller.java:    private static void deployPlugin(final UpdateSite.Plugin plugin)
blackduck-installer-plugin/src/main/java/com/blackducksoftware/integration/jenkins/BDInstaller.java:        plugin.deploy(true);
cloudbees-enterprise-plugins-plugin/src/main/java/com/cloudbees/jenkins/plugins/enterpriseplugins/PluginImpl.java:            require("cloudbees-aws-deployer","1.13"),
cloudbees-enterprise-plugins-plugin/src/main/java/com/cloudbees/jenkins/plugins/enterpriseplugins/PluginImpl.java:                                p.deploy().get();
cloudbees-enterprise-plugins-plugin/src/main/java/com/cloudbees/jenkins/plugins/enterpriseplugins/PluginImpl.java:                            p.deploy().get();
recipe-plugin/src/main/java/org/jenkinsci/plugins/recipe/ingredients/PluginIngredient.java:                        installSource.deploy(true).get();
recipe-plugin/src/main/java/org/jenkinsci/plugins/recipe/ingredients/PluginIngredient.java:                            installSource.deploy(true).get();
system-config-dsl-plugin/src/main/java/org/jenkinsci/plugins/systemconfigdsl/Root.java:            jenkins.getUpdateCenter().getPlugin(name).deploy(true).get();
thin-backup-plugin/src/main/java/org/jvnet/hudson/plugins/thinbackup/restore/HudsonRestore.java:              return plugin.deploy();
```

does not indicate any such calls. This sort of search is of course imprecise since an `UpdateSite` object might have been obtained via method chaining without the type being imported, but the only way I can see this happening is via getters from an `UpdateCenter`, which is turn can only be obtained via a `getUpdateCenter` method that would also match.

oleg-nenashev reviewed (dismissed):
Needs clarification about `deploy()` binary compatibility and logging level changes before merging

jglick commented:
Cryptic CI failures…
